### PR TITLE
Fix summary generation when empty strings are present.

### LINKED ISSUE #2339: Route summary can include empty strings
If ways with no names end up being the largest part of the route, then the summary may end as an odd-looking string like:

`, M-30`

In this case, the first summary item are a collection of ways with no names.

Example route against http://router.project-osrm.org/route/v1/car/-3.663499999999999,40.4679;-3.7053999999999974,40.4067?steps=true

We should exclude all empty road names from summary calculation.

### PR DESCRIPTION
**Pull request by danpat**
Also corrects step ordering when steps have equal durations.

Fixes #2339

### CODE REVIEW

File: include/engine/guidance/assemble_leg.hpp [hunk 1]
```diff
@@ -34,14 +34,22 @@ struct NamedSegment
...
 
 template <std::size_t SegmentNumber>
+
 std::array<std::uint32_t, SegmentNumber> summarizeRoute(const std::vector<PathData> &route_data)
 {
     // merges segments with same name id
     const auto collapse_segments = [](std::vector<NamedSegment> &segments)
     {
         auto out = segments.begin();
         auto end = segments.end();
```
Comment (MoKob) [inline include/engine/guidance/assemble_leg.hpp]: `const`

File: features/testbot/summary.feature [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        And the ways
+            | nodes | name |
+            | ab    | ab   |
+            | bc    | bc   |
+            | cd    | cd   |
+            | de    | de   |
+            | ef    |      |
+
+        When I route I should get
+            | f    | a  | ,de,cd,bc,ab,ab   | de, bc   |
```
Comment (MoKob) [inline features/testbot/summary.feature]: missing header?

**Comment by MoKob** [comment]
I added another testcase and fixed one of yours. After the fixing (added header), cucumber fails. Please also review my testcase.

File: include/engine/guidance/assemble_leg.hpp [hunk 3]
```diff
@@ -54,7 +62,8 @@ std::array<std::uint32_t, SegmentNumber> summarizeRoute(const std::vector<PathDa
                 *out = *in;
             }
         }
-        return out;
+        BOOST_ASSERT(out != end);
+        return ++out;
```
Comment (daniel-j-h) [inline include/engine/guidance/assemble_leg.hpp]: Is this `std::unique(first, last, sameNameid)` ?

Comment (MoKob) [inline include/engine/guidance/assemble_leg.hpp]: close. It accumulates duration and does something else. the something else seems like it should be unique, but it does not work. I don't know exactly what it is supposed to be.

File: features/testbot/summary.feature [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            | nodes | name |
+            | ab    | ab   |
+            | bc    | bc   |
+            | cd    |      |
+            | de    | de   |
+            | df    | df   |
+
+        When I route I should get
+            | from | to | route          | summary  |
+            | e    | a  | de,,bc,ab   | de, bc   |
```
Comment (TheMarex) [inline features/testbot/summary.feature]: This returns a route that is named:

> de,,ab,ab 

Note no `bc`. But the s ummary is named `de,bc`. It seems the post-processing changes the name of the `bc` segment. @MoKob can you take a look at this tomorrow?

Comment (MoKob) [inline features/testbot/summary.feature]: It's not a postprocessing issue but a generation issue. The part of the name announcement has an error for unnamed-roads. Fixing it and adding test-cases.

Comment (MoKob) [inline features/testbot/summary.feature]: Well, this is not really a bug, at least not in post processing. It is how to announce name-changes with no-name.

I added https://github.com/Project-OSRM/osrm-backend/pull/2367 to handle this situation a bit better.

In addition, the string needs to be `de,,bc,ab,ab` (currently ,,->bc is not announced, will change https://github.com/Project-OSRM/osrm-backend/pull/2367)

**Pull request merged**